### PR TITLE
[Tabs] Add Selection Indicator

### LINKED ISSUE #7744: [Tabs] Internal issue: b/135609974
This was filed as an internal issue. If you are a Googler, please visit [b/135609974](http://b/135609974) for more details.

---
#### Internal data
- Associated internal bug: [b/135609974](http://b/135609974)
- Blocked by: https://github.com/material-components/material-components-ios/issues/7644
- Blocked by: https://github.com/material-components/material-components-ios/issues/7643

### PR DESCRIPTION
**Pull request by romoore**
Adds selection indicator support and rendering to MDCTabBarView. The indicator animates between items and is sized to the full width of the item. Theming (color) of the indicator will be done in a follow-up change.

|Scrollable without Safe Area Insets|Justified with Safe Area Insets|
|---|---|
|tabs-si-nosafe|tabs-si-safe|

Closes #7744

### CODE REVIEW

File: components/Tabs/src/TabBarView/MDCTabBarView.h [hunk 1]
```diff
@@ -36,6 +37,11 @@ __attribute__((objc_subclassing_restricted)) @interface MDCTabBarView : UIScroll
 /** The tab bar view delegate. */
 @property(nullable, nonatomic, weak) id<MDCTabBarViewDelegate> tabBarDelegate;
 
+/**
+ The template for the selection indicator.
+ */
+@property(nonnull, nonatomic, strong) id<MDCTabBarViewIndicatorTemplate> selectionIndicatorTemplate;
```
Comment (galiak11) [inline components/Tabs/src/TabBarView/MDCTabBarView.h:44]: Nit: Could be easier to understand if had a short explanation of what template means in this context (unless already explained elsewhere?)

File: components/Tabs/src/TabBarView/MDCTabBarView.m [hunk 2]
```diff
@@ -181,6 +214,7 @@ - (void)setSelectedItem:(UITabBarItem *)selectedItem animated:(BOOL)animated {
     [self updateTitleColorForAllViews];
     [self updateImageTintColorForAllViews];
     [self updateTitleFontForAllViews];
+    [self didSelectItemAtIndex:NSNotFound animateTransition:animated];
```
Comment (galiak11) [inline components/Tabs/src/TabBarView/MDCTabBarView.m:206]: Nit: What does selecting an NSNotFound item do?

File: components/Tabs/src/TabBarView/MDCTabBarView.m [hunk 3]
```diff
@@ -603,4 +647,114 @@ - (void)didTapItemView:(UITapGestureRecognizer *)tap {
   }
 }
 
+- (void)applySelectionTemplateToSelectionViewForItemView:(UIView *)itemView {
+  if (!itemView) {
+    return;
```
Comment (galiak11) [inline components/Tabs/src/TabBarView/MDCTabBarView.m:641]: When would this be called with a nil view? 
Is that a likely scenario? 
Nit - If not, did you consider using a parameter assert?

File: components/Tabs/src/TabBarView/MDCTabBarView.m [hunk 4]
```diff
@@ -603,4 +647,114 @@ - (void)didTapItemView:(UITapGestureRecognizer *)tap {
...
+  // Ask the template for attributes.
+  id<MDCTabBarViewIndicatorTemplate> template = self.selectionIndicatorTemplate;
+  MDCTabBarViewIndicatorAttributes *indicatorAttributes =
+      [template indicatorAttributesForContext:context];
+
+  // Update the selection indicator.
+  [self.selectionIndicatorView applySelectionIndicatorAttributes:indicatorAttributes];
+}
+
+- (void)didSelectItemAtIndex:(NSUInteger)index animateTransition:(BOOL)animate {
```
Comment (galiak11) [inline components/Tabs/src/TabBarView/MDCTabBarView.m:721]: Nit - Can you add a comment explaining this method? Specifically what happens when the index is NotFound...

**Review by fluxminds — APPROVED** [review]

**Review by galiak11 — APPROVED** [review]

File: components/Tabs/src/TabBarView/MDCTabBarView.m (hunk 2 above)
Comment (romoore) [reply components/Tabs/src/TabBarView/MDCTabBarView.m:206]: It makes nothing selected.

File: components/Tabs/src/TabBarView/MDCTabBarView.m (hunk 3 above)
Comment (romoore) [reply components/Tabs/src/TabBarView/MDCTabBarView.m:641]: At the moment, I don't believe it is. It's just a bit of defensive coding to keep the code safer for the future. I don't think it should be an error if the user passes `nil` (the parameter is "null unspecified").  There may be a legitimate case of unselecting or "animating out" the selection indicator of an item is removed.

File: components/Tabs/src/TabBarView/MDCTabBarView.m (hunk 4 above)
Comment (romoore) [reply components/Tabs/src/TabBarView/MDCTabBarView.m:721]: Done.

File: components/Tabs/src/TabBarView/MDCTabBarView.h (hunk 1 above)
Comment (romoore) [reply components/Tabs/src/TabBarView/MDCTabBarView.h:44]: Added a bit more. If we need further details, it should probably be in a follow-up PR.

**Pull request merged**